.config/emacs/config.org
#+TITLE: Emacs config
#+AUTHOR: yakate
#+OPTIONS: toc:2

This is my emacs config, using org, cuz it's cool :p

disable errors. only because of evil-collection
#+begin_src emacs-lisp
(setq warning-minimum-level :emergency)
#+end_src

* Settings
** scrollbar
#+begin_src emacs-lisp
  (scroll-bar-mode -1)
#+end_src
** toolbar
#+begin_src emacs-lisp
  (tool-bar-mode -1)
#+end_src
** menu bar

#+begin_src emacs-lisp
  (menu-bar-mode -1)
#+end_src
** backups
Both variables are set to nil, because they really make a mess in my filesystem.
#+begin_src emacs-lisp
  (setq make-backup-files nil)
  (setq auto-save-default nil)
  (setq create-lockfiles nil)
#+end_src
** shorten stuff
#+begin_src emacs-lisp
  (defalias 'yes-or-no-p 'y-or-n-p)
#+end_src
** line numbers
#+begin_src emacs-lisp
(global-display-line-numbers-mode 1)
(global-visual-line-mode t)
(setq display-line-numbers-type 'relative )
#+end_src
** scrolling
getting rid of the annoying page jumps
#+begin_src emacs-lisp
  (setq scroll-conservatively 101)
#+end_src
* Fonts
#+begin_src emacs-lisp
(add-to-list 'default-frame-alist
             '(font . "JetBrainsMono Nerd Font-12"))
#+end_src
* Functions
Some useful functions I've stolen from people.

** dt's emacs reload function:
#+begin_src emacs-lisp
  (defun reload-init-file ()
  (interactive)
  (load-file user-init-file)
  (load-file user-init-file))

#+end_src
* Package Manager

- melpa and use-package
#+begin_src emacs-lisp
  (require 'package)
  (setq package-enable-at-startup nil)

  (add-to-list 'package-archives
               '("melpa" . "https://melpa.org/packages/"))
  (package-initialize)

  (unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package)
  (setq use-package-always-ensure t))
#+end_src
* Themes
Doom emacs default themes:
#+begin_src emacs-lisp
    (use-package doom-themes
    :ensure t
    :config
    (setq doom-themes-enable-bold t    ; if nil, bold is universally disabled
          doom-themes-enable-italic t) ; if nil, italics is universally disabled
    (doom-themes-neotree-config)
    (doom-themes-org-config))
#+end_src

ef themes:
#+begin_src emacs-lisp
(use-package ef-themes :ensure t)
#+end_src

loading the theme:
#+begin_src emacs-lisp
  (load-theme 'doom-one t)
  ;(load-theme 'ef-winter t)
#+end_src
* Keybindings
Yes, this is technicaly a package, but I've separated it just because it's important
#+begin_src emacs-lisp
  (use-package general
    :ensure t
    :config
    (general-evil-setup)

    (general-create-definer leader
      :states '(normal insert visual emacs)
      :keymaps 'override
      :prefix "SPC" 
      :global-prefix "M-SPC") 

    (leader
      ;;better M-x
      "SPC" '(counsel-M-x :wk "counsel M-x")
      ;;coding related keybindings
      "TAB TAB" '(comment-line :wk "comment a line")
      ;;config related keybindings
      "c" '(:ignore t :wk "config")
      "cc" '((lambda () (interactive) (find-file "~/.emacs.d/config.org")) :wk "edit emacs conf")
      "cr" '(reload-init-file :wk "Reload conf")
      "ct" '(load-theme :wk "load a theme")

      ;;buffer related keybindings
      "b" '(:ignore t :wk "buffer")
      "bb" '(switch-to-buffer :wk "Switch buffer")
      "bk" '(kill-this-buffer :wk "Kill this buffer")
      "bn" '(next-buffer :wk "Next buffer")
      "bp" '(previous-buffer :wk "Previous buffer")
      "br" '(revert-buffer :wk "Reload buffer")

      ;;file related keybindings
      "f" '(:ignore t :wk "file")
      "ff" '(counsel-find-file :wk "Find a file")
      "fe" '(dired :wk "emacs file explorer (dired)")

      ;; terminal related keybindings
      "t" '(:ignore t :wk "terminals")
      "tt" '(vterm-toggle :wk "toggle vterm")
      "te" '(eshell :wk "eshell")

      ;;dashboard
      "h" '(:ignore t :wk "dashboard")
      "hh" '(dashboard-open :wk "open dashboard")

      ;;toggling stuff
      "p" '(:ignore t :wk "toggle")
      "pl" '(display-line-numbers-mode :wk "toggle line numbers")
      "pt" '(centaur-tabs-mode :wk "toggle centaur tabs")
      "pf" '(global-flycheck-mode :wk "toggle flycheck")
      "pn" '(neotree-toggle :wk "toggle neotree")

      "v" '(:ignore t :wk "code")
      "vc" '(comment-line :wk "comment current line"))
  )

#+end_src
* Packages
** evil
#+begin_src emacs-lisp
  (use-package evil 
             :demand t
             :init
             (evil-mode))
(use-package evil-collection
  :after (evil)
  :init
  (setq evil-collection-key-blacklist '("SPC")) ;;blacklisting space key, because of general
  :config
  (evil-collection-init))
#+end_src
** ivy and counsel

Install ivy, counsel and related utilities
#+begin_src emacs-lisp

  (use-package counsel
    :config (counsel-mode))

  (use-package ivy
    :bind
    ;; ivy-resume resumes the last Ivy-based completion.
    (("C-c C-r" . ivy-resume)
     ("C-x B" . ivy-switch-buffer-other-window))
    :custom
    (setq ivy-use-virtual-buffers t)
    (setq ivy-count-format "(%d/%d) ")
    (setq enable-recursive-minibuffers t)
    :config
    (ivy-mode))

  (use-package all-the-icons-ivy-rich :ensure t
    :init (all-the-icons-ivy-rich-mode 1))

  (use-package ivy-rich
    :after ivy
    :init (ivy-rich-mode 1) ;; this gets us descriptions in M-x.
    :custom
    (ivy-virtual-abbreviate 'full
     ivy-rich-switch-buffer-align-virtual-buffer t
     ivy-rich-path-style 'abbrev)
    :config
    (ivy-set-display-transformer 'ivy-switch-buffer
                                 'ivy-rich-switch-buffer-transformer))

  (use-package ivy-posframe
    :config
    (setq ivy-posframe-display-functions-alist '((t . ivy-posframe-display)))
    (setq ivy-posframe-display-functions-alist '((t . ivy-posframe-display-at-frame-top-center)))
    (ivy-posframe-mode 1))

#+end_src
** which-key
(my beloved)
#+begin_src emacs-lisp
  (use-package which-key
    :config
    (setq which-key-side-window-location 'bottom
          which-key-separator "   "
          which-key-add-column-padding 1
          which-key-max-display-columns nil
          which-key-min-display-lines 5
          which-key-side-window-slot 10
          which-key-idle-delay 0.2
          which-key-allow-imprecise-window-fit nil)
  
    (which-key-mode 1))

#+end_src
** company
Emacs code completion framework
#+begin_src emacs-lisp
   ;   (use-package company
   ;   :defer t
   ;   :config
   ;   (add-hook 'after-init-hook 'global-company-mode)
   ;   :init
   ;   (global-company-mode 1))
#+end_src
** auto-complete
#+begin_src emacs-lisp 

(use-package auto-complete
  :ensure t
  :config
  (ac-config-default))
  #+end_src
** dashboard
#+begin_src emacs-lisp
  (use-package dashboard
    :diminish
    :init
    (setq dashboard-center-content t) 
    (setq dashboard-startup-banner 'logo)
    (setq dashboard-banner-logo-title "Emacs > vim")
    (dashboard-open)
    :config
    (dashboard-setup-startup-hook))
#+end_src
** modeline
default doom emacs modeline
#+begin_src emacs-lisp
  (use-package doom-modeline
    :init (doom-modeline-mode 1)
    :config
    (setq doom-modeline-height 30
          doom-modeline-bar-width 5
          doom-modeline-presp-name t
          doom-modeline-presp-icon t))

#+end_src
** ido
#+begin_src emacs-lisp
  (use-package ido-vertical-mode
  :config
  (setq ido-enable-flex-matching t)
  (setq ido-everywhere t)
  (setq ido-vertical-define-keys 'C-n-and-C-p-only)      
  :init
  (ido-mode 1)
  (ido-vertical-mode 1))
#+end_src
** eye candy
These are some of the less 'necessary' packages, that just look good.
#+begin_src emacs-lisp

    (use-package all-the-icons)
    (use-package page-break-lines)
    (use-package rainbow-mode
        :hook org-mode prog-mode)
#+end_src
** projectile
#+begin_src emacs-lisp
  (use-package projectile
     :config
     (projectile-mode 1))
#+end_src
** shells and terminals
Should be shell and term, really. :p
#+begin_src emacs-lisp
  ;; vterm
  (use-package vterm
    :config
    (setq shell-file-name "/bin/zsh"
        vterm-max-scrollback 5000))


  ;; vterm-toggle
  (use-package vterm-toggle
    :after vterm
    :config
    (setq vterm-toggle-fullscreen-p nil)
    (setq vterm-toggle-scope 'project)
    (add-to-list 'display-buffer-alist
                 '((lambda (buffer-or-name _)
                       (let ((buffer (get-buffer buffer-or-name)))
                         (with-current-buffer buffer
                           (or (equal major-mode 'vterm-mode)
                               (string-prefix-p vterm-buffer-name (buffer-name buffer))))))
                    (display-buffer-reuse-window display-buffer-at-bottom)
                    (reusable-frames . visible)
                    (window-height . 0.3))))

  ;;eshell syntax
  (use-package eshell-syntax-highlighting
    :after esh-mode
    :config
    (eshell-syntax-highlighting-global-mode 1))


#+end_src
** flycheck
#+begin_src emacs-lisp
  (use-package flycheck
    :defer t
    :init (global-flycheck-mode))
#+end_src
** org stuff
everything related to org, not really much to say here. It's all just the more useless eye candy :p.
#+begin_src emacs-lisp
  ;;org-bullets
  (add-hook 'org-mode-hook 'org-indent-mode)
  (use-package org-bullets)
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
  ;; org-tempo
  (require 'org-tempo)
  ;; toc-org
  (use-package toc-org
      :commands toc-org-enable
      :init (add-hook 'org-mode-hook 'toc-org-enable))
#+end_src
** beacon

#+begin_src emacs-lisp
  (use-package beacon
  :config
     (beacon-mode 1))
  (global-hl-line-mode t)
#+end_src
** centaur tabs
#+begin_src emacs-lisp
(use-package centaur-tabs
  :config
  (setq centaur-tabs-style "bar")
  (setq centaur-tabs-height 32)
  (setq centaur-tabs-set-icons t)
  (setq centaur-tabs-gray-out-icons 'buffer)
  (setq centaur-tabs-set-bar 'under)
  (setq x-underline-at-descent-line t)
  ;;enabling tabs on startup doesn't really make sense, so I'm only gonna toggle them via the keybindings
  (centaur-tabs-mode -1))

#+end_src
** language support

#+begin_src emacs-lisp
(use-package lua-mode)
(use-package haskell-mode)
(use-package rust-mode)
#+end_src
** neotree
#+begin_src emacs-lisp
(use-package neotree :ensure t
  :config
  (setq neo-theme (if (display-graphic-p) 'icons 'arrow)))
#+end_src
** dired icons
#+begin_src emacs-lisp
  (use-package all-the-icons-dired)
#+end_src
** magit
#+begin_src emacs-lisp
  (use-package magit :ensure t)
#+end_src


.config/emacs/config.el
(defvar elpaca-installer-version 0.6)
(defvar elpaca-directory (expand-file-name "elpaca/" user-emacs-directory))
(defvar elpaca-builds-directory (expand-file-name "builds/" elpaca-directory))
(defvar elpaca-repos-directory (expand-file-name "repos/" elpaca-directory))
(defvar elpaca-order '(elpaca :repo "https://github.com/progfolio/elpaca.git"
                              :ref nil
                              :files (:defaults "elpaca-test.el" (:exclude "extensions"))
                              :build (:not elpaca--activate-package)))
(let* ((repo  (expand-file-name "elpaca/" elpaca-repos-directory))
       (build (expand-file-name "elpaca/" elpaca-builds-directory))
       (order (cdr elpaca-order))
       (default-directory repo))
  (add-to-list 'load-path (if (file-exists-p build) build repo))
  (unless (file-exists-p repo)
    (make-directory repo t)
    (when (< emacs-major-version 28) (require 'subr-x))
    (condition-case-unless-debug err
        (if-let ((buffer (pop-to-buffer-same-window "*elpaca-bootstrap*"))
                 ((zerop (call-process "git" nil buffer t "clone"
                                       (plist-get order :repo) repo)))
                 ((zerop (call-process "git" nil buffer t "checkout"
                                       (or (plist-get order :ref) "--"))))
                 (emacs (concat invocation-directory invocation-name))
                 ((zerop (call-process emacs nil buffer nil "-Q" "-L" "." "--batch"
                                       "--eval" "(byte-recompile-directory \".\" 0 'force)")))
                 ((require 'elpaca))
                 ((elpaca-generate-autoloads "elpaca" repo)))
            (progn (message "%s" (buffer-string)) (kill-buffer buffer))
          (error "%s" (with-current-buffer buffer (buffer-string))))
      ((error) (warn "%s" err) (delete-directory repo 'recursive))))
  (unless (require 'elpaca-autoloads nil t)
    (require 'elpaca)
    (elpaca-generate-autoloads "elpaca" repo)
    (load "./elpaca-autoloads")))
(add-hook 'after-init-hook #'elpaca-process-queues)
(elpaca `(,@elpaca-order))

;; Install use-package support
(elpaca elpaca-use-package
  ;; Enable :elpaca use-package keyword.
  (elpaca-use-package-mode)
  ;; Assume :elpaca t unless otherwise specified.
  (setq elpaca-use-package-by-default t))

;; Block until current queue processed.
(elpaca-wait)

;;When installing a package which modifies a form used at the top-level
;;(e.g. a package which adds a use-package key word),
;;use `elpaca-wait' to block until that package has been installed/configured.
;;For example:
;;(use-package general :demand t)
;;(elpaca-wait)

;; Expands to: (elpaca evil (use-package evil :demand t))
(use-package evil
    :init      ;; tweak evil's configuration before loading it
    (setq evil-want-integration t) ;; This is optional since it's already set to t by default.
    (setq evil-want-keybinding nil)
    (setq evil-vsplit-window-right t)
    (setq evil-split-window-below t)
    (evil-mode))
  (use-package evil-collection
    :after evil
    :config
    (setq evil-collection-mode-list '(dashboard dired ibuffer))
    (evil-collection-init))
  (use-package evil-tutor)

;;Turns off elpaca-use-package-mode current declaration
;;Note this will cause the declaration to be interpreted immediately (not deferred).
;;Useful for configuring built-in emacs features.
(use-package emacs :elpaca nil :config (setq ring-bell-function #'ignore))

;; Don't install anything. Defer execution of BODY
(elpaca nil (message "deferred"))

(use-package general
   :config
   (general-evil-setup)
   (general-create-definer my/leader-keys
   :states '(normal insert visual emacs)
   :keymaps 'override
   :prefix "SPC" ;; set leader
   :global-prefix "M-SPC") 

 (my/leader-keys
    "b"  '(:ignore t :wk "buffer")
    "bb" '(switch-to-buffer :wk "Switch buffer")
    "bk" '(kill-this-buffer :wk "Kill this buffer")
    "bn" '(next-buffer :wk "Next buffer")
    "bp" '(previous-buffer :wk "Previous buffer")
    "br" '(revert-buffer :wk "Reload buffer")
    "bt" '(treemacs :wk "open treemacs")
    "be" '(eshell :wk "open eshell")
    
    )

)

(use-package treemacs
  :ensure t
  :defer t
  :init
  (with-eval-after-load 'winum
    (define-key winum-keymap (kbd "M-0") #'treemacs-select-window))
  :config
  (progn
    (setq treemacs-collapse-dirs                   (if treemacs-python-executable 3 0)
          treemacs-deferred-git-apply-delay        0.5
          treemacs-directory-name-transformer      #'identity
          treemacs-display-in-side-window          t
          treemacs-eldoc-display                   'simple
          treemacs-file-event-delay                2000
          treemacs-file-extension-regex            treemacs-last-period-regex-value
          treemacs-file-follow-delay               0.2
          treemacs-file-name-transformer           #'identity
          treemacs-follow-after-init               t
          treemacs-expand-after-init               t
          treemacs-find-workspace-method           'find-for-file-or-pick-first
          treemacs-git-command-pipe                ""
          treemacs-goto-tag-strategy               'refetch-index
          treemacs-header-scroll-indicators        '(nil . "^^^^^^")
          treemacs-hide-dot-git-directory          t
          treemacs-indentation                     2
          treemacs-indentation-string              " "
          treemacs-is-never-other-window           nil
          treemacs-max-git-entries                 5000
          treemacs-missing-project-action          'ask
          treemacs-move-forward-on-expand          nil
          treemacs-no-png-images                   nil
          treemacs-no-delete-other-windows         t
          treemacs-project-follow-cleanup          nil
          treemacs-persist-file                    (expand-file-name ".cache/treemacs-persist" user-emacs-directory)
          treemacs-position                        'left
          treemacs-read-string-input               'from-child-frame
          treemacs-recenter-distance               0.1
          treemacs-recenter-after-file-follow      nil
          treemacs-recenter-after-tag-follow       nil
          treemacs-recenter-after-project-jump     'always
          treemacs-recenter-after-project-expand   'on-distance
          treemacs-litter-directories              '("/node_modules" "/.venv" "/.cask")
          treemacs-project-follow-into-home        nil
          treemacs-show-cursor                     nil
          treemacs-show-hidden-files               t
          treemacs-silent-filewatch                nil
          treemacs-silent-refresh                  nil
          treemacs-sorting                         'alphabetic-asc
          treemacs-select-when-already-in-treemacs 'move-back
          treemacs-space-between-root-nodes        t
          treemacs-tag-follow-cleanup              t
          treemacs-tag-follow-delay                1.5
          treemacs-text-scale                      nil
          treemacs-user-mode-line-format           nil
          treemacs-user-header-line-format         nil
          treemacs-wide-toggle-width               70
          treemacs-width                           35
          treemacs-width-increment                 1
          treemacs-width-is-initially-locked       t
          treemacs-workspace-switch-cleanup        nil)

    ;; The default width and height of the icons is 22 pixels. If you are
    ;; using a Hi-DPI display, uncomment this to double the icon size.
    ;;(treemacs-resize-icons 44)

    (treemacs-follow-mode t)
    (treemacs-filewatch-mode t)
    (treemacs-fringe-indicator-mode 'always)
    (when treemacs-python-executable
      (treemacs-git-commit-diff-mode t))

    (pcase (cons (not (null (executable-find "git")))
                 (not (null treemacs-python-executable)))
      (`(t . t)
       (treemacs-git-mode 'deferred))
      (`(t . _)
       (treemacs-git-mode 'simple)))

    (treemacs-hide-gitignored-files-mode nil))
  :bind
  (:map global-map
        ("M-0"       . treemacs-select-window)
        ("C-x t 1"   . treemacs-delete-other-windows)
        ("C-x t t"   . treemacs)
        ("C-x t d"   . treemacs-select-directory)
        ("C-x t B"   . treemacs-bookmark)
        ("C-x t C-t" . treemacs-find-file)
        ("C-x t M-t" . treemacs-find-tag)))

(use-package treemacs-evil
  :after (treemacs evil)
  :ensure t)

(use-package treemacs-projectile
  :after (treemacs projectile)
  :ensure t)

(use-package treemacs-icons-dired
  :hook (dired-mode . treemacs-icons-dired-enable-once)
  :ensure t)

(use-package treemacs-magit
  :after (treemacs magit)
  :ensure t)

(use-package treemacs-persp ;;treemacs-perspective if you use perspective.el vs. persp-mode
  :after (treemacs persp-mode) ;;or perspective vs. persp-mode
  :ensure t
  :config (treemacs-set-scope-type 'Perspectives))

(use-package treemacs-tab-bar ;;treemacs-tab-bar if you use tab-bar-mode
  :after (treemacs)
  :ensure t
  :config (treemacs-set-scope-type 'Tabs))

;(use-package treemacs-evil)

(use-package eshell-syntax-highlighting
  :after esh-mode
  :config
  (eshell-syntax-highlighting-global-mode +1))

;; eshell-syntax-highlighting -- adds fish/zsh-like syntax highlighting.
;; eshell-rc-script -- your profile for eshell; like a bashrc for eshell.
;; eshell-aliases-file -- sets an aliases file for the eshell.
  
(setq eshell-rc-script (concat user-emacs-directory "eshell/profile")
      eshell-aliases-file (concat user-emacs-directory "eshell/aliases")
      eshell-history-size 5000
      eshell-buffer-maximum-lines 5000
      eshell-hist-ignoredups t
      eshell-scroll-to-bottom-on-input t
      eshell-destroy-buffer-when-process-dies t
      eshell-visual-commands'("bash" "fish" "htop" "ssh" "top" "zsh"))

(use-package beacon
:ensure t
:init
:config
(beacon-mode 1))

(use-package auto-complete
:ensure t
:init
(progn
   (ac-config-default)
   (global-auto-complete-mode t)
 ))

(use-package which-key
  :init
    (which-key-mode 1)
  :config
  (setq which-key-side-window-location 'bottom
	  which-key-sort-order #'which-key-key-order-alpha
	  which-key-sort-uppercase-first nil
	  which-key-add-column-padding 1
	  which-key-max-display-columns nil
	  which-key-min-display-lines 6
	  which-key-side-window-slot -10
	  which-key-side-window-max-height 0.25
	  which-key-idle-delay 0.8
	  which-key-max-description-length 25
	  which-key-allow-imprecise-window-fit t
	  which-key-separator " → " ))

(use-package rainbow-mode
:hook org-mode prog-mode)

(add-to-list 'custom-theme-load-path "~/.config/emacs/themes/")

;;uncomment one of the following themes to use it
;(load-theme 'darkmacs t)
;(load-theme 'lightmacs t)
;(load-theme 'dracula t)
;(load-theme 'solarized t)
;(load-theme 'gruvbox-dark-medium t)
(load-theme 'charcoal t)

;;uncomment the following line to get a scratchpad instead of the tutorial screen.
;;(setq inhibit-startup-message t)
;;-------------------------------------------------------------------------------------------------------------
(use-package dashboard
  :ensure t 
  :init
  (setq initial-buffer-choice 'dashboard-open)
  (setq dashboard-set-heading-icons t)
  (setq dashboard-set-file-icons t)
  (setq dashboard-banner-logo-title "Yqll")
  (setq dashboard-startup-banner 'logo) ;; use standard emacs logo as banner
  ;(setq dashboard-startup-banner "/home/yakate/programming/oyakate.github.io/img/rosepfp2.jpg")  ;; use custom image as banner
  (setq dashboard-center-content nil) ;; set to 't' for centered content
  (setq dashboard-items '((recents . 5)
                         (agenda . 5 )
                         (bookmarks . 3)
                         (projects . 3)
                         (registers . 3)))
  :custom
  (dashboard-modify-heading-icons '((recents . "file-text")
                                    (bookmarks . "book")))
  :config
  (dashboard-setup-startup-hook))

;;---------------------------------------------------------------------------------------------------------------

(tool-bar-mode -1)

(scroll-bar-mode -1)

(set-face-attribute 'default nil
  :font "JetBrains Mono Nerd Font"
  :height 110
  :weight 'medium)

(global-display-line-numbers-mode 1)
(global-visual-line-mode t)

(use-package projectile
  :config
  (projectile-mode 1))

(add-hook 'org-mode-hook 'org-indent-mode)
(use-package org-bullets)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))

(use-package toc-org
    :commands toc-org-enable
    :init (add-hook 'org-mode-hook 'toc-org-enable))
